### PR TITLE
[ML] Make ManagedLedger storage configurable

### PR DESCRIPTION
**Pull request by sijie**
*Motivation*

This is the first step to allow supporting different storage implementations for Pulsar

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     * @return the stats provider.
+     */
+    StatsProvider getStatsProvider();
+
+    /**
+     * Return the default bookkeeper client.
+     *
+     * @return the default bookkeeper client.
+     */
+    BookKeeper getBookKeeperClient();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:69]: If we want to support different storage, why the interface contains the bookkeeper client related method, seems this should be handled in the implementation?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Initialize the managed ledger storage.
+     *
+     * @param conf service config
+     * @param zkClient zk client
+     * @param bookkeperProvider bookkeeper provider
+     * @throws Exception
+     */
+    void initialize(ServiceConfiguration conf,
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:46]: If we wrap an object or abstract a client API, which has nothing to do with the storage medium, would it be better? 
If it is abstract, we don’t have to use zk or bk.
In the future, we will increase the storage type, only need to modify this object, and no need to modify the interface.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java (hunk 1 above)
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:69]: Probably this interface should be org.apache.bookkeeper.client.api.Bookkeeper (the new BK API) if we want to look forward for the future.
I know that this is not possible with the current codebase, that is bound to the old API.
So this comment is not a request to change but just a reminder for the future.

**Review by eolivelli — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java (hunk 1 above)
Comment (sijie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:69]: @codelipenghui This is the first attempt to provide an interface. It is a bit difficult to hide everything at this moment. Hence I marked the interface as private and unstable. So we can iterate changes over this interface.

Comment (sijie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:69]: @eolivelli see reasons above.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java (hunk 2 above)
Comment (sijie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:46]: @315157973 see my comment in https://github.com/apache/pulsar/pull/9397#discussion_r568467892

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java (hunk 1 above)
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:69]: sure @sijie
I had already approved this patch.
please go ahead

Comment (sijie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/storage/ManagedLedgerStorage.java:69]: Need approval from @codelipenghui and @315157973

**Review by zymap — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Review by 315157973 — APPROVED** [review]

**Review by jiazhai — APPROVED** [review]

**Pull request merged**